src/utilities/valueToLiteral.js

```diff
@@ -0,0 +1,97 @@
+import { inspect } from '../jsutils/inspect';
+import { isIterableObject } from '../jsutils/isIterableObject';
+
+import type { ConstValueNode } from '../language/ast';
+import { Kind } from '../language/kinds';
+
+import type { GraphQLInputType } from '../type/definition';
+import {
+  getNamedType,
+  isLeafType,
+  isInputObjectType,
+} from '../type/definition';
+
+/**
+ * Produces a GraphQL Value AST given a JavaScript value.
+ *
+ * A GraphQL type may be provided, which will be used to interpret different
+ * JavaScript values if it defines a `valueToLiteral` method.
+ *
+ * | JavaScript Value  | GraphQL Value        |
+ * | ----------------- | -------------------- |
+ * | Object            | Input Object         |
+ * | Array             | List                 |
+ * | Boolean           | Boolean              |
+ * | String            | String               |
+ * | Number            | Int / Float          |
+ * | null / undefined  | Null                 |
+ *
+ * Note: This function does not perform any type validation or coercion.
+ */
+export function valueToLiteral(
```

Comment:
I think this method tries to be too clever: the type argument should never be undefined and the whole conversion logic should be based on the type.

If it is a List type => take the value apart and perform it per value.
If Object type => iterate over field definitions and call it per field
If Scalar/Enum => call the `valueToLiteral` on its type definition.

If you have an input object as type and the value is a boolean it should raise an error: at the moment it returns a `boolean` literal which is wrong.

Reply:
I thought about this, but the problem with this approach is that we're operating on "external" pre-coerced, pre-validated values and it's entirely possible that those values do not match the types. These conversion functions are used before validation takes place in #3049. That may require us to define validation rules again in each of these functions, ideally they can be passed through so that we define validation once.

Reply:
We also need reasonable defaults for when we don't have a type, since that's going to be common for any custom scalars with complex definitions (like the JSON type)

Reply:
Took me a bit to wrap my head around, but I think our misunderstanding comes from a fundamental principles about   `Scalar.literalToValue`/`Scalar.valueToLiteral` does: I think these two methods should follow the same principles as `parseValue/parseLiteral` and should throw errors if something is wrong. I want basically another "coercing" (already a bit overloaded term btw) which converts from Value and Literal and vice verse. 

Looking at the `ID` implementation in this PR I see that exactly this is missing: currently it returns `undefined` instead of throwing `GraphQLError`. Building on this principle I think method is basically the equivalent of `coerceInputValueImpl` but going from `value` to `literal` .

The reason I want this: for example for a JSON `literalToValue` method : you would want to throw a clear error message for an expected Literal deep inside the provided `literal`. What is the alternative? Return `undefined`? Silently ignore?

There is also general value in having a well defined, validated literal or external value for inputs. 

And as you mentioned: this would mean validated twice, yes. But I don't see logic defined twice really: all validation logic resided inside the Scalar except for Input Objects. And I am sure we can extract/generalize this logic if that is a concern.

To be specific: we check for two things when validating input objects: if all required fields are provided and not more fields are provided.

Reply:
I don't necessarily disagree with that. I think being more strict is useful, and I'm sure we could find a way to generalize the validation portion.

The root of the issue is still variables within custom scalars. We have no types there and all bets are off. As long as we decide to preserve that behavior we need untyped transform

Reply:
I don't think I follow completely regarding the variables:

We actually have a type of the variables and we could use that type to convert the External input value into an Ast Literal in a well defined manner. We could then simply replace the VariableReference ast node with the Ast Literal we got. 

Custom Scalar implementation can't type check variables values anyway at the moment: they just get a variable and they basically take the value. 

I think that would preserver the current behavior.